pd/src/app.rs

```diff
@@ -443,24 +452,63 @@ impl App {
 
                 let mut next_rates = Vec::new();
                 let mut next_validator_statuses = Vec::new();
+
+                // this is a bit complicated: because we're in the EndBlock phase, and the
+                // delegations in this block have not yet been committed, we have to combine
+                // the delegations in pending_block with the ones already committed to the
+                // state. otherwise the delegations committed in the epoch threshold block
+                // would be lost.
+                let mut committed_delegation_changes =
+                    state.delegation_changes(prev_epoch.index).await?;
+                for (id_key, delta) in &pending_block.lock().unwrap().delegation_changes {
+                    let _ = *committed_delegation_changes
+                        .entry(id_key.clone())
+                        .and_modify(|change| *change += delta)
+                        .or_insert(*delta);
+                }
+
                 for current_rate in &current_rates {
                     let identity_key = current_rate.identity_key.clone();
 
                     let funding_streams = state.funding_streams(identity_key.clone()).await?;
 
                     let next_rate = current_rate.next(&next_base_rate, funding_streams);
 
-                    let delegation_token_supply = state
+                    // TODO: if a validator isn't part of the consensus set, should we ignore them
+                    // and not update their rates?
+                    let delegation_delta = committed_delegation_changes
+                        .get(&identity_key)
+                        .unwrap_or(&0i64);
+
+                    // update staking token supply
+                    let unbonded_amount =
+                        current_rate.unbonded_amount(delegation_delta.abs() as u64);
```

Comment:
is this correct? should we be updating delegation token and staking token supply based on current_rate, or should we be using next_rate?

Reply:
Yes, exactly, using current_rate here is correct, we crossed the epoch boundary at lines 409-413 above:
```rust
                // We've finished processing the last block of `epoch`, so we've
                // crossed the epoch boundary, and (prev | current | next) are:
                let prev_epoch = epoch;
                let current_epoch = prev_epoch.next();
                let next_epoch = current_epoch.next();
```
All the (un)delegations we're processing were prepared in the (now-previous) epoch using that epoch's `next_rate`, which is now `current_rate`.